homeassistant/components/shelly/config_flow.py

```diff
@@ -48,33 +52,69 @@ async def async_step_user(self, user_input=None):
+                self.host = user_input[CONF_HOST]
+                info = await self._async_get_info(self.host)
             except HTTP_CONNECT_ERRORS:
                 errors["base"] = "cannot_connect"
             except Exception:  # pylint: disable=broad-except
                 _LOGGER.exception("Unexpected exception")
                 errors["base"] = "unknown"
 
             else:
+                await self.async_set_unique_id(info["mac"])
+                self._abort_if_unique_id_configured()
                 if info["auth"]:
-                    return self.async_abort(reason="auth_not_supported")
+                    return await self.async_step_credentials()
 
                 try:
-                    device_info = await validate_input(self.hass, user_input)
+                    device_info = await validate_input(self.hass, self.host, user_input)
                 except HTTP_CONNECT_ERRORS:
                     errors["base"] = "cannot_connect"
                 except Exception:  # pylint: disable=broad-except
                     _LOGGER.exception("Unexpected exception")
                     errors["base"] = "unknown"
                 else:
-                    await self.async_set_unique_id(device_info["mac"])
                     return self.async_create_entry(
-                        title=device_info["title"] or user_input["host"],
+                        title=device_info["title"] or self.host,
                         data=user_input,
                     )
 
         return self.async_show_form(
-            step_id="user", data_schema=DATA_SCHEMA, errors=errors
+            step_id="user", data_schema=HOST_SCHEMA, errors=errors
+        )
+
+    async def async_step_credentials(self, user_input=None):
+        """Handle the credentials step."""
+        errors = {}
+        schema = None
+        if user_input is not None:
+            try:
+                device_info = await validate_input(self.hass, self.host, user_input)
+            except aiohttp.ClientResponseError as error:
+                if error.status == 401:
+                    errors["base"] = "invalid_auth"
+                else:
+                    errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(
+                    title=device_info["title"] or self.host,
+                    data={**user_input, **{CONF_HOST: self.host}},
+                )
+            schema = vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME, default=user_input[CONF_USERNAME]): str,
+                    vol.Required(CONF_PASSWORD, default=user_input[CONF_PASSWORD]): str,
+                }
+            )
+
+        if schema is None:
+            schema = vol.Schema({CONF_USERNAME: str, CONF_PASSWORD: str})
```

Comment:
Should these schema items really be optional?

Reply:
Hmm, I thought the default was required. I guess this also applies to `HOST_SCHEMA` as well, then.

Reply:
The default for dict items is optional.

Reply:
If the items should always be required in this schema, we can define it one time and use `dict.get` to look up available user input to set as default values.

Reply:
I thought about this too, but the dict is `None` where there's no user input. I'll figure something out.